compiler/src/regular-grammar/index.ts

```diff
@@ -283,75 +354,291 @@ class RegularGrammar extends CompilerClass {
+  /**
+   * Constructs a Deterministic Finite Automaton (DFA) from a Non-deterministic Finite Automaton (NFA).
+   * 1. Create a DFA start state from the NFA start state
+   * 2. For each DFA state and each input symbol:
+   *    a. Compute the set of NFA states reachable from the current DFA state
+   *    b. If this set of states is new, add it as a new DFA state
+   * 3. Repeat step 2 until no new DFA states are created
+   * 4. Mark DFA states as final if they contain any NFA final states
+   **/
+
+  toDFA() {
+    const nfa: FAGraph = this.toEpsilonFreeFA().result;
+    const dfa: FAGraph = {};
+    const alphabet: Set<string> = new Set();
+    const stateQueue: Set<string>[] = [];
+    const processedStates: Set<string>[] = [];
+    const DEAD_STATE = 'Φ';
+
+    // Get the alphabet(List of terminals)
+    for (const state in nfa) {
+      for (const symbol in nfa[state].nodes) {
+        alphabet.add(symbol);
+      }
+    }
+
+    dfa[DEAD_STATE] = { nodes: {}, final: false };
+    for (const symbol of alphabet) {
+      dfa[DEAD_STATE].nodes[symbol] = new Set([DEAD_STATE]);
+    }
+
+    // Helper function to get the next state set
+    const getNextStateSet = (
+      currentSet: Set<string>,
+      symbol: string
+    ): Set<string> => {
+      const nextSet = new Set<string>();
+      for (const state of currentSet) {
+        if (nfa[state].nodes[symbol]) {
+          nfa[state].nodes[symbol].forEach((nextState) =>
+            nextSet.add(nextState)
+          );
+        }
+      }
+      return nextSet;
+    };
+
+    // Helper function to convert set to sorted string (for state naming)
+    const setToString = (set: Set<string>): string =>
+      Array.from(set).sort().join(',');
+
+    // Start with the initial state
+    const initialState = new Set(['S']);
+    stateQueue.push(initialState);
+
+    while (stateQueue.length > 0) {
+      const currentSet = stateQueue.shift()!;
+      const currentState = setToString(currentSet);
+
+      if (processedStates.some((set) => setToString(set) === currentState))
+        continue;
+
+      processedStates.push(currentSet);
+
+      dfa[currentState] = { nodes: {}, final: false };
+
+      // Check if the current state set contains any final state from NFA
+      dfa[currentState].final = Array.from(currentSet).some(
+        (state) => nfa[state].final
+      );
+
+      for (const symbol of alphabet) {
+        const nextSet = getNextStateSet(currentSet, symbol);
+        if (nextSet.size > 0) {
+          const nextState = setToString(nextSet);
+          dfa[currentState].nodes[symbol] = new Set([nextState]);
+
+          if (!processedStates.some((set) => setToString(set) === nextState)) {
+            stateQueue.push(nextSet);
+          }
+        } else {
+          // Transition to dead state if next set is empty
+          dfa[currentState].nodes[symbol] = new Set([DEAD_STATE]);
+        }
+      }
+    }
+
+    this.result = dfa;
+    return this;
+  }
+  minimizeDFA() {
```

Comment:
Let's add the algorithm used here as comments. I was finding it hard to understand it so any reference links would also help here.